// @lerna-lite/cli/dist/cli-commands/cli-run-commands.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const filter_options_1 = require("../filter-options");
/**
 * @see https://github.com/yargs/yargs/blob/master/docs/advanced.md#providing-a-command-module
 */
exports.default = {
    command: 'run <script>',
    describe: 'Run an npm script in each package that contains that script',
    builder: (yargs) => {
        yargs
            .example('$0 run build -- --silent', '# `npm run build --silent` in all packages with a build script')
            .parserConfiguration({
            'populate--': true,
        })
            .positional('script', {
            describe: 'The npm script to run. Pass flags to send to the npm client after --',
            type: 'string',
        })
            .options({
            'cmd-dry-run': {
                group: 'Command Options:',
                describe: 'Displays the process command that would be performed without executing it.',
                type: 'boolean',
            },
            'npm-client': {
                group: 'Command Options:',
                describe: 'Executable used to run scripts (npm, yarn, pnpm, ...).',
                defaultDescription: 'npm',
                type: 'string',
                requiresArg: true,
            },
            stream: {
                group: 'Command Options:',
                describe: 'Stream output with lines prefixed by package.',
                type: 'boolean',
            },
            parallel: {
                group: 'Command Options:',
                describe: 'Run script with unlimited concurrency, streaming prefixed output.',
                type: 'boolean',
            },
            'no-bail': {
                group: 'Command Options:',
                describe: 'Continue running script despite non-zero exit in a given package.',
                type: 'boolean',
            },
            bail: {
                // proxy for --no-bail
                hidden: true,
                type: 'boolean',
            },
            // This option controls prefix for stream output so that it can be disabled to be friendly
            // to tools like Visual Studio Code to highlight the raw results
            'no-prefix': {
                group: 'Command Options:',
                describe: 'Do not prefix streaming output.',
                type: 'boolean',
            },
            prefix: {
                // proxy for --no-prefix
                hidden: true,
                type: 'boolean',
            },
            profile: {
                group: 'Command Options:',
                describe: 'Profile script executions and output performance profile to default location.',
                type: 'boolean',
            },
            'profile-location': {
                group: 'Command Options:',
                describe: 'Output performance profile to custom location instead of default project root.',
                type: 'string',
            },
            'skip-nx-cache': {
                hidden: true,
                type: 'boolean',
            },
        });
        return (0, filter_options_1.filterOptions)(yargs);
    },
    handler: async (argv) => {
        try {
            // @ts-ignore
            // eslint-disable-next-line import/no-unresolved
            const { RunCommand } = await Promise.resolve().then(() => __importStar(require('@lerna-lite/run')));
            new RunCommand(argv);
        }
        catch (err) {
            console.error(`"@lerna-lite/run" is optional and was not found. Please install it with "npm install @lerna-lite/run -D -W".`, err);
        }
    },
};
